litestar/_openapi/plugin.py

```diff
@@ -43,13 +49,52 @@ def _build_openapi_schema(self) -> OpenAPI:
 
     def provide_openapi(self) -> OpenAPI:
         if not self._openapi_schema:
-            self._openapi_schema = self._build_openapi_schema()
+            self._openapi_schema = self._build_openapi()
         return self._openapi_schema
 
+    @staticmethod
+    def provide_openapi_schema(__openapi: Any) -> Dict[str, Any]:  # noqa: UP006
+        return __openapi.to_schema()  # type: ignore[no-any-return]
+
+    def create_openapi_router(self) -> Router:
+        def create_handler(plugin: OpenAPIRenderPlugin) -> HTTPRouteHandler:
+            paths = [plugin.path] if isinstance(plugin.path, str) else list(plugin.path)
+            if any(path.endswith(self.openapi_config.root_schema_site) for path in paths):
+                paths.append("/")
+
+            @get(paths, media_type=plugin.media_type, sync_to_thread=False)
+            def _handler(request: Request, __openapi_schema: Dict[str, Any]) -> bytes:  # noqa: UP006
+                return plugin.render(request, __openapi_schema)
+
+            return _handler
+
+        router = Router(
+            self.openapi_config.path or "/schema",
+            route_handlers=[create_handler(plugin) for plugin in self.openapi_config.render_plugins],
+            include_in_schema=False,
+            dto=None,
+            return_dto=None,
+        )
+        for plugin in self.openapi_config.render_plugins:
+            plugin.receive_router(router)
+
+        return router
+
     def on_app_init(self, app_config: AppConfig) -> AppConfig:
         if app_config.openapi_config:
             self._openapi_config = app_config.openapi_config
-            app_config.route_handlers.append(self.openapi_config.openapi_controller)
+            if (controller := app_config.openapi_config.openapi_controller) is not None:
```

Comment:
If a controller is passed to the config, then a user is sub-classing the `OpenAPIController` and so we register that and don't do anything different. This maintains the backward compat with the controller pattern.